zebra/zebra_gr.c

```diff
@@ -183,10 +317,420 @@ zebra_gr_client_reconnect(struct zserv *client)
+			client->gr_instance_count);
+
+		if ((info->gr_enable) &&
+				(client->gr_instance_count > 0))
+			client->gr_instance_count--;
+
+		zebra_gr_client_info_delte(client, info);
+		break;
+	case ZEBRA_CLIENT_GR_CAPABILITIES:
+		/* Allocate bgp info */
+		if (!info)
+			info = zebra_gr_client_info_create(client);
+
+		/* Udpate other parameters */
+		if (!info->gr_enable) {
+			client->gr_instance_count++;
+
+			LOG_GR("%s: Cient %s GR enabled count %d",
+				__func__, zebra_route_string(client->proto),
+				client->gr_instance_count);
+
+			info->capabilities = api->cap;
+			info->stale_removal_time =
+					api->stale_removal_time;
+			info->vrf_id = api->vrf_id;
+			info->gr_enable = true;
+		}
+		break;
+	case ZEBRA_CLIENT_RIB_STALE_TIME:
+		LOG_GR("%s: Client %s stale time update event",
+			__func__, zebra_route_string(client->proto));
+
+		/* Update the stale removal timer */
+		if (info && info->t_stale_removal == NULL) {
+
+			LOG_GR("%s: Stale time: %d is now update to: %d",
+				__func__, info->stale_removal_time,
+				api->stale_removal_time);
+
+			info->stale_removal_time =
+					api->stale_removal_time;
+		}
+
+			break;
+	case ZEBRA_CLIENT_ROUTE_UPDATE_COMPLETE:
+		LOG_GR(
+		    "%s: Client %s route update completed for AFI %d, SAFI %d"
+			 , __func__, zebra_route_string(client->proto),
+			 api->afi, api->safi);
+		if (info)
+			info->route_sync[api->afi][api->safi] = true;
+		break;
+	case ZEBRA_CLIENT_ROUTE_UPDATE_PENDING:
+		LOG_GR("%s: Client %s route update pending for AFI %d, SAFI %d",
+			__func__, zebra_route_string(client->proto),
+			api->afi, api->safi);
+		if (info)
+			info->af_enabled[api->afi][api->safi] = true;
+		break;
+	default:
+		break;
+	}
+}
+
+/*
+ * Handler for capabilities that are received from client.
+ */
+static void
+zebra_client_capabilties_handler(struct zserv *client,
+				 struct zapi_cap *api)
+{
+	switch (api->cap) {
+	case ZEBRA_CLIENT_GR_CAPABILITIES:
+	case ZEBRA_CLIENT_ROUTE_UPDATE_PENDING:
+	case ZEBRA_CLIENT_GR_DISABLE:
+	case ZEBRA_CLIENT_RIB_STALE_TIME:
+		/*
+		 * For all the cases we need to update the client info.
+		 */
+		zebra_client_update_info(client, api);
+		break;
+	case ZEBRA_CLIENT_ROUTE_UPDATE_COMPLETE:
+		/*
+		 * After client info has been updated delete all
+		 * stale routes
+		 */
+		zebra_client_update_info(client, api);
+		zebra_gr_process_client_stale_routes(client, api->vrf_id);
+		break;
+	default:
+		break;
+	}
+}
+
+/*
+ * Function to decode and call appropriate functions
+ * to handle client capabilities.
+ */
+void
+zread_client_capabilities(ZAPI_HANDLER_ARGS)
+{
+	struct zapi_cap api;
+	struct stream *s;
 
+	s = msg;
+
+	if (zapi_capabilities_decode(s, &api)) {
+		LOG_GR("%s: Error in reading capabilities for client %s",
+			    __func__,  zebra_route_string(client->proto));
+		return;
+	}
+
+	switch (client->proto) {
+	case ZEBRA_ROUTE_BGP:
+		/* For now only BGP is being handled. */
+		zebra_client_capabilties_handler(client, &api);
+		break;
+	default:
```

Comment:
> This makes no sense to me. What is even remotely specific about bgp here? Why limit this way? Plus fix the spelling in the zebra_client_capabilities_handler function. That will drive me crazy

Check was added only with one intention that few of the capability would be sent only from BGP and clients like OSPF does not sent it. For example in case of OSPF there is no EOR and hence that capability would not be sent. I will remove the check as if OSPF does not send then it would not exercise that part of code. It was a safe check which I will remove it.